File: backendV2/src/modules/user/dto/update-profile.dto.ts
import {
  IsBoolean,
  IsNumber,
  IsOptional,
  IsString,
  Matches,
} from 'class-validator';

export class UpdateProfileDto {
  @IsOptional()
  @IsString()
  full_name?: string;

  @IsOptional()
  @Matches(/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/, {
    message: 'Email không hợp lệ',
  })
  email?: string;

  @IsOptional()
  @IsString()
  @Matches(/^\+?[0-9]\d{1,14}$/, { message: 'Số điện thoại không hợp lệ' })
  phone_number?: string;

  @IsOptional()
  @IsString()
  avatar?: string;

  @IsOptional()
  @IsString()
  cccd?: string;

  @IsOptional()
  @IsBoolean({ message: 'is_active phải là boolean' })
  is_active?: boolean;

  @IsOptional()
  @IsString()
  license_number?: string;

  @IsOptional()
  @IsString()
  address?: string;

  @IsOptional()
  @IsNumber()
  lat?: number;

  @IsOptional()
  @IsNumber()
  lng?: number;

  // @IsOptional()
  // @IsObject({ message: 'Địa chỉ phải là object' })
  // address?: Partial<Address>;
}

File: mobile_app_config/lib/features/profile/data/models/update_profile_dto.dart
/// DTO for updating user profile
class UpdateProfileDto {
  final String? fullName;
  final String? phoneNumber;
  final String? email;
  final String? cccd;
  final String? avatar;
  final String? address;
  final double? lat;
  final double? lng;

  UpdateProfileDto({
    this.fullName,
    this.phoneNumber,
    this.email,
    this.cccd,
    this.avatar,
    this.address,
    this.lat,
    this.lng,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};

    if (fullName != null && fullName!.isNotEmpty) {
      data['full_name'] = fullName;
    }
    if (phoneNumber != null && phoneNumber!.isNotEmpty) {
      data['phone_number'] = phoneNumber;
    }
    if (email != null && email!.isNotEmpty) {
      data['email'] = email;
    }
    if (cccd != null && cccd!.isNotEmpty) {
      data['cccd'] = cccd;
    }
    if (avatar != null && avatar!.isNotEmpty) {
      data['avatar'] = avatar;
    }
    if (address != null && address!.isNotEmpty) {
      data['address'] = address;
    }
    if (lat != null) {
      data['lat'] = lat;
    }
    if (lng != null) {
      data['lng'] = lng;
    }

    return data;
  }
}
